test_echo.py

```diff
@@ -1,31 +1,103 @@
+        'HTTP/1.1 200 OK\r\nContent-Type: text/plain\r\nContent-length: 18\r\n\r\neverything is okay'
+
+
+def test_response_error():
+    """Test that function returns appropriate byte string."""
+    from echo_server import response_error
+    assert response_error(404, 'Not Found') == 'HTTP/1.1 404 Not Found\r\nContent-Type: text/plain\r\nContent-length: 18\r\n\r\nNot Found'
+
+
+def test_simple_parse_request():
+    """Test that function returns appropriate URI."""
+    from echo_server import parse_request
+    assert parse_request('GET /path/to/index.html HTTP/1.1\r\nHost: www.host1.com:80\r\n\r\n') == \
+        '/path/to/index.html'
+
+
+def test_put_parse_request():
+    """Test that function returns error when PUT method requested."""
+    from echo_server import parse_request
+    assert parse_request('PUT /path/to/index.html HTTP/1.1\r\nHost: www.host1.com:80\r\n') == \
+        'HTTP/1.1 405 Method Not Allowed\r\nContent-Type: text/plain\r\nContent-length: 18\r\n\r\nMethod Not Allowed'
+
+
+def test_HTTP10_parse_request():
+    """Test that function returns error on HTTP/1.0 request."""
+    from echo_server import parse_request
+    assert parse_request('GET /path/to/index.html HTTP/1.0\r\nHost: www.host1.com:80\r\n') == \
+        'HTTP/1.1 505 HTTP Version Not Supported\r\nContent-Type: text/plain\r\nContent-length: 18\r\n\r\nHTTP Version Not Supported'
+
+
+def test_HTTP20_parse_request():
+    """Test that function returns error on HTTP/2.0 request."""
+    from echo_server import parse_request
+    assert parse_request('GET /path/to/index.html HTTP/2.0\r\nHost: www.host1.com:80\r\n') == \
+        'HTTP/1.1 505 HTTP Version Not Supported\r\nContent-Type: text/plain\r\nContent-length: 18\r\n\r\nHTTP Version Not Supported'
+
+
+def test_server():
+    """In response to incoming request, server responds with 200 ok message."""
+    from echo_server import response_ok
+    assert client('GET /path/to/index.html HTTP/1.1\r\nHost: www.host1.com:80\r\n\r\n') == response_ok()
+
+
+def test_server_505():
+    """In response to HTTP20 request, server responds with error."""
+    assert client('GET /path/to/index.html HTTP/2.0\r\nnHost: www.host1.com:80\r\n\r\n') == \
+        'HTTP/1.1 505 HTTP Version Not Supported\r\nContent-Type: text/plain\r\nContent-length: 18\r\n\r\nHTTP Version Not Supported'
+
+
+def test_server_405():
+    """In response to HTTP20 request, server responds with error."""
+    assert client('PUT /path/to/index.html HTTP/1.1\r\nHost: www.host1.com:80\r\n\r\n') == \
+        'HTTP/1.1 405 Method Not Allowed\r\nContent-Type: text/plain\r\nContent-length: 18\r\n\r\nMethod Not Allowed'
+
+
+def test_server_505_2():
+    """In response to HTTP20 SET request, server responds with error."""
+    assert client('SET /path/to/index.html HTTP/2.0\r\nHost: www.host1.com:80\r\n\r\n') == \
+        'HTTP/1.1 505 HTTP Version Not Supported\r\nContent-Type: text/plain\r\nContent-length: 18\r\n\r\nHTTP Version Not Supported'
```

Comment:
A reasonable selection of tests, both functional and unit.  Lines of code are ridiculously over-long.  How can you cut down the length while still testing the same things (or better things, even).